### PR TITLE
feat: typescript entrypoint

### PR DESCRIPTION
**Pull request by TomChv**
Handle dagger do <cmd> with string arguments.
Split code into multiples files and function.
Add local import and test documentation.

Fixes DAG-1423

### CODE REVIEW

File: sdk/nodejs/example/simple-entrypoint/package.json
Comment (sipsma) [inline sdk/nodejs/example/simple-entrypoint/package.json:8]: This is totally fine, but I'm curious if you think every `start` script users write will be exactly this? If so, could we just put `node --loader ts-node/esm ./main.mts` in the typescript runtime on [this line](https://github.com/dagger/dagger/pull/5305/files#diff-53e3c4d2665984237716f1e4be37edc950485b8cf2c89d9e89230ced7f342cfdR24)?

File: sdk/nodejs/server/convertor.ts
Comment (sipsma) [inline sdk/nodejs/server/convertor.ts:82]:
```suggestion
export function entrypointsMetadatatoGQLSchema(
```

File: sdk/nodejs/server/server.ts
Comment (sipsma) [inline sdk/nodejs/server/server.ts:31]: Do you think we need to get every file in the dir here, or can we just provide the file that has the `serveCommands` call (which should be in the callstack of this function)?

My guess was that `program.getSourceFiles` would then handle visiting every file imported by that one, but could be wrong.

Not a big deal, just a potential small simplification I am curious about.

**Review by sipsma — COMMENTED** [review]
Looks really great, thanks for putting this together so quickly! Have a couple minor comments/questions, but really the only thing I think we should do before merging is add some integ tests for what's working so far.

I added a bunch of test utils recently, you can see the latest in this PR: https://github.com/dagger/dagger/pull/5308/files#diff-b6e594474025769dd7b1a4bfbeb8b43ab3a26b34796bb76f76be91a0cbbd410d

I'm hoping to get that merged soon, but in the meantime you can rebase on top of that one or just write the tests separately, whatever you prefer.

File: sdk/nodejs/example/simple-entrypoint/package.json
Comment (TomChv) [reply sdk/nodejs/example/simple-entrypoint/package.json:8]: It could be this or something else, by using the script `start`, we let the user choose how to start the ci.
This way, he may rename his file `ci.ts` or `ci.mts` or use a different loader or anything. What we actually need is to launch the script, we can let the user define its own way.

**Comment by TomChv** [comment]
> I added a bunch of test utils recently, you can see the latest in this PR: https://github.com/dagger/dagger/pull/5308/files#diff-b6e594474025769dd7b1a4bfbeb8b43ab3a26b34796bb76f76be91a0cbbd410d

Nice! I might add more tests in a further PR then!

**Comment by helderco** [comment]
There's a lot of unrelated commits here, I assume you did a merge from main instead of rebase. Can you fix that?

**Comment by TomChv** [comment]
I'll, my bad!

File: sdk/nodejs/server/server.ts
Comment (slumbering) [inline sdk/nodejs/server/server.ts:6]:
```suggestion
import { entrypointsMetadatatoGQLSchema } from "./convertor.js"
```

File: sdk/nodejs/server/server.ts
Comment (slumbering) [inline sdk/nodejs/server/server.ts:87]:
```suggestion
  const gqlSchema = entrypointsMetadatatoGQLSchema(metadatas)
```

File: demos/typescript-entrypoints/main.mts
Comment (slumbering) [inline demos/typescript-entrypoints/main.mts:1]: Should be rename to index.mts to follow our conventions.

File: sdk/nodejs/server/server.ts
Comment (slumbering) [inline sdk/nodejs/server/server.ts:12]: should ...args be optional since you're not using with compute for instance ?
Here: https://github.com/dagger/dagger/pull/5305/files#diff-5ed95a013d71c623984de86737f366350b9a6a82e1b2e17e11b310da9cc5060bR9

File: sdk/nodejs/server/server.ts
Comment (slumbering) [inline sdk/nodejs/server/server.ts:104]: We should probably document serveCommands since it will be displayed in the generated docs.

File: sdk/nodejs/server/server.ts
Comment (TomChv) [reply sdk/nodejs/server/server.ts:12]: Nope, args is a destructured so it's actually optionnal because there can be 0, one or as many args as you want :)

**Review by slumbering — APPROVED** [review]
LGTM 🚢

**Comment by TomChv** [comment]
@sipsma Waiting for your approval to merge :)

**Review by sipsma — APPROVED** [review]
LGTM! I just merged the CLI PR, which includes the updates to the integ test tooling you may want here: https://github.com/dagger/dagger/pull/5308

I do think it's extremely important to get test coverage for TS in there ASAP. You can see what @helderco is doing with integrating python here for a good example on how to approach this: https://github.com/dagger/dagger/pull/5300

However, no need for that to be in this PR if you'd prefer a follow up.

**Comment by sipsma** [comment]
@TomChv I just merged https://github.com/dagger/dagger/pull/5360 which creates a kind of painful rebase for you here, so I'm going to push a commit to your PR here with the updates to save you some time, don't worry about making them yourself

**Comment by TomChv** [comment]
Ok it seems that dependencies installation is a bit messed up with npm but it works well with `yarn`, the PR is almost ready!
All bugs are almost fixed!

**Comment by TomChv** [comment]
@sipsma It seems the test `TestImportedProjectDir` cannot be performed the same way on the Typescript SDK because the `fs.readdirSync` is relative to the path of the file who executes the code, not the path where the command is executed.
Should I update my command to check from the project dir instead of the workdir? That would change the whole purpose of the test.

Do you have any idea of what we could do to be consistent between all tests?

**Comment by gerhard** [comment]
What would help move this one along?

**Comment by helderco** [comment]
@gerhard, it's been on hold because of the shift in Zenith design and major refactor from @sipsma:
- https://github.com/dagger/dagger/pull/5443

More context in [Discord](https://discord.com/channels/707636530424053791/1129130729469657218/1129130742371319929).

**Comment by TomChv** [comment]
Closing this PR for now since Zenith include too much breaking change that invalidate this PR.
However some work could be used later to integrate entrypoint on the Node SDK!

**Pull request closed without merging**
